cmd/containerd-shim-runhcs-v1/exec_wcow_podsandbox.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/containerd/containerd/runtime/v2/task"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func newWcowPodSandboxExec(ctx context.Context, events publisher, tid, bundle string) *wcowPodSandboxExec {
+	logrus.WithFields(logrus.Fields{
+		"tid": tid,
+		"eid": "",
+	}).Debug("newWcowPodSandboxExec")
+
+	wpse := &wcowPodSandboxExec{
+		events:     events,
+		tid:        tid,
+		bundle:     bundle,
+		state:      shimExecStateCreated,
+		exitStatus: 255, // By design for non-exited process status.
+		exited:     make(chan struct{}),
+	}
+	return wpse
+}
+
+var _ = (shimExec)(&wcowPodSandboxExec{})
+
+// wcowPodSandboxExec is a special exec type that actually holds no real
+// resources. The WCOW model has two services the HCS/HNS that actually hold
+// open any shared namespace outside of the container/process itself. So we
+// don't actually need to create them in the case of a POD for the sandbox
+// container.
+//
+// Note: This is only true today because CRI defines an API that allows for no
+// query/state/exec/io operations against the POD sandbox container itself. If
+// this changes we need to rethink this.
+//
+// Note 2: This is only true today because Windows contains only a shared
+// `NetworkNamespace` controlled via CNI in the CRI layer and held open by the
+// HNS and not by any container runtime attribute. If we ever have a shared
+// namespace that requires a container we will have to rethink this.
+type wcowPodSandboxExec struct {
+	events publisher
+	// tid is the task id of the container hosting this process.
+	//
+	// This MUST be treated as read only in the lifetime of the exec.
+	tid string
+	// bundle is the on disk path to the folder containing the `process.json`
+	// describing this process. If `id==""` the process is described in the
+	// `config.json`.
+	//
+	// This MUST be treated as read only in the lifetime of the exec.
+	bundle string
+
+	// sl is the state lock that MUST be held to safely read/write any of the
+	// following members.
+	sl         sync.Mutex
+	state      shimExecState
+	pid        int
+	exitStatus uint32
+	exitedAt   time.Time
+
+	// exited is a wait block which waits async for the process to exit.
+	exited chan struct{}
+}
+
+func (wpse *wcowPodSandboxExec) ID() string {
+	return ""
+}
+
+func (wpse *wcowPodSandboxExec) Pid() int {
+	wpse.sl.Lock()
+	defer wpse.sl.Unlock()
+	return wpse.pid
+}
+
+func (wpse *wcowPodSandboxExec) State() shimExecState {
+	wpse.sl.Lock()
+	defer wpse.sl.Unlock()
+	return wpse.state
+}
+
+func (wpse *wcowPodSandboxExec) Status() *task.StateResponse {
+	wpse.sl.Lock()
+	defer wpse.sl.Unlock()
+
+	var s containerd_v1_types.Status
+	switch wpse.state {
+	case shimExecStateCreated:
+		s = containerd_v1_types.StatusCreated
+	case shimExecStateRunning:
+		s = containerd_v1_types.StatusRunning
+	case shimExecStateExited:
+		s = containerd_v1_types.StatusStopped
+	}
+
+	return &task.StateResponse{
+		ID:         "",
```

Comment:
@crosbymichael - https://github.com/containerd/containerd/blob/master/runtime/v2/task/shim.proto#L92 is this `ID` in StateResponse the `ConatinerID` or `ExecID`. Does this need s a similar fix that I made for `TaskDelete` event?

Reply:
If anything, I would day `ContainerID` as you get a state request for a specific process by ID, so that request should return with the `ID` equal to the same.

Reply:
Ok fair enough. I will make it the `ContainerID` but I am also going to contribute a change to add `ExecID` its non breaking and will be "more correct"

Reply:
I mean, keep ID, but add `ContainerID` in addition

Reply:
I don't want to break backwards compat with these changes

Reply:
https://github.com/containerd/containerd/pull/3016